Jsonparser step3

AppleCEO opened the pull request:
요구사항 구현했습니다.

AppleCEO committed:
생성 - GrammarChecker

AppleCEO committed:
구현 - GrammarChecker 구조체 일부 구현

괄호 열고 닫히는 것 스택으로 확인하는 bracketCheck 메서드 구현
오브젝트 안에 배열이나 오브젝트가 나오는 것 확인하는 getObjectList, objectExist, arrayExist 메서드 구현

AppleCEO committed:
구현 - GrammarChecker 구조체 구현 완료

배열 안에 오브젝트가 있는 경우 각 오브젝트를 검증하는 eachGrammarCheck 메서드 추가

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    static private func arrayCheck (_ input: String) -> Bool {
+        if input.matches("\\[[A-z\"0-9., ]+\\]") {
```
여기서 사용하는 RegEx 문법 문자열도 하드 코딩하지말고 어떤 형식이든 코드로 표현하세요.

AppleCEO replied:
제가 좋아하는 enum 타입으로 구현해봤습니다.

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    static private func eachGrammarCheck (_ input: String) throws -> String {
+        
```
이 함수는 왜 한 줄이 떨어져 있는건가요? 어떤 의도에서 띄운건가요?

AppleCEO replied:
실수였습니다.

godrm commented on JSONParser/JSONParser/JsonParser.swift:
```diff
@@ -11,7 +11,7 @@ import Foundation
     
 
-    static func makeJson (_ inputSplited: [String]) -> [JsonType] {
+    static func makeJson (_ inputSplited: [String]) throws -> [JsonType] {
```
이 함수는 Json 데이터를 만드는 걸까요? Json 데이터를 Parsing 하는걸까요?
Json을 만든다는 것로 해석하게 되는데 헷갈리지 않을까요?

AppleCEO replied:
메서드명 변경하였습니다.

godrm commented on JSONParser/JSONParser/main.swift:
```diff
@@ -12,10 +12,10 @@ func main () {
         inputView.readInput()
         do {
+            try GrammarChecker.grammarCheck(inputView.valueEntered)
             json = try Converter.inputToJson(inputView.valueEntered)
```
GrammarChecker.grammarCheck()에 대해 try - catch 가 Converter.inputToJson() 동작 try - catch 와 같은 맥락일까요?
try - catch를 하고나면 항상 흐름제어가 있어야 하는데 두 가지가 같은 흐름을 갖게 될까요? 
지금 아래 catch 에서 흐름제어가 없고 정상적인 경우와 마찬가지로 동작하고 있습니다. 
이 부분을 꼭 구분하세요.

AppleCEO replied:
GrammarChecker.grammarCheck() 와 Converter.inputToJson 는 둘 다 input 에 문제가 있다면 오류를 출력하기 때문에 같은 try - catch 구문에 넣어봤습니다.
두 함수가 정상적으로 실행되야지 아래에 break 가 실행되면서 아래로 내려가집니다. 그러한 점에서 흐름제어가 정상적으로 되고 있다고 생각합니다.

godrm replied:
예외가 발생해서 catch가 되고 나면 어디로 진행하게 되는지 궁금했었습니다.

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+    
+    static private func eachGrammarCheck (_ input: String) throws -> String {
+        
+        if !(objectCheck(input) || arrayCheck(input)) && input != " ]" {
```
여기 마지막 `input != " ]"` 비교식은 왜 필요한건가요?

AppleCEO replied:
[ { "name" : "KIM JUNG", "alias" : "JK", "level" : 5, "married" : true }, 
{ "name" : "YOON JISU", "alias" : "crong", "level" : 4, "married" : true } ] 
라고 입력이 된다면 
"[ { "name" : "KIM JUNG", "alias" : "JK", "level" : 5, "married" : true }" 와 
"{ "name" : "YOON JISU", "alias" : "crong", "level" : 4, "married" : true }" 와 
" ]" 로 잘려져서 문법 검사를 시작합니다. 이러한 로직 때문에 정상적인 상황임에도 불구하고 " ]" 를 문법 검사하다가 오류가 발생하였습니다. 이러한 상황이 발생하지 않도록 막아놓았습니다.

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+        let curryBracketOpen = DevideCharacter.curlyBracketOpen
+        let curryBracketClose = DevideCharacter.curlyBracketClose
+        let squareBracketOpen = DevideCharacter.squareBracketOpen
+        let squareBracketClose = DevideCharacter.squareBracketClose
```
여기 왜 다시 let으로 다시 선언하고 있는건가요?
심지어 `DevideCharacter.curlyBracketOpen`을 `curryBracketOpen` 이라고 오타가 있는데요. 이게 어떤 의도인가요?

AppleCEO replied:
rawValue 를 사용하기 위해 let 으로 선언하고 있습니다.

godrm replied:
아래에서 비교하는 character 타입이 DevideCharacter 라면 let 선언없이 .curlyBrackerOpen 이랑 비교할 수 있지 않나요?

AppleCEO replied:
선언없이 비교하도록 개선하였습니다.

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+        
+        for character in input {
+            eachCheck += String(character)
+            if character == curryBracketOpen.rawValue || character == squareBracketOpen.rawValue {
```
curryBracketOpen 같은 타입을 선언했으면 rawValue 와 직접 비교하지말고 타입으로 비교할 수 있도록 표현해보세요.

AppleCEO replied:
타입으로 비교하도록 수정했습니다.

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+        
+        for object in getObjectList(input)
+        {
+            if objectCheck(object) || arrayCheck(object) {
```
brackerCheck(), objectCheck(), arrayCheck() 함수 이름보다 구체적인걸로 표현하세요.
지금 함수 이름은 동사가 뒤에 있고, -check 만으로는 무엇을 검증하려는건지 명확하지 않습니다.

AppleCEO replied:
불명확한 메서드명들을 전체 수정했습니다.

godrm commented on JSONParser/JSONParser/Converter.swift:
```diff
@@ -12,27 +12,19 @@ struct Converter {
     static func inputToJson (_ valueEntered: String) throws -> [JsonType] {
```
inputToJson() 메소드 이름도 개선해보세요. Input을 붙이지 않는게 좋습니다.
Converter는 입력을 받은 것만 호출해서 사용한다고 생각하면 안됩니다. 입력을 받던, 파일에서 읽어오던, 네트워크에서 값을 받아도 Converter를 사용할 수도 있는거니까요.

AppleCEO replied:
stringToJson으로 변경해봤습니다!

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    static private func objectCheck (_ input: String) -> Bool {
+        if input.matches("\\{[A-z\"0-9:., ]+\\}") {
```
여기 objectCheck()와 arrayCheck() 에 있는 정규표현식으로 Object 와 Array는 완벽하게 매칭이 되나요? 어떻게 검증해봤나요? 테스트 코드라도 있을까요?

AppleCEO replied:
테스트코드 작성해봤습니다!

godrm reviewed (commented):
정규표현식을 어떻게 검증해봤나요? 
가장 좋은 방법은 단위 테스트 코드를 만들어서 테스트로 검증하는 겁니다. 
다음 피드백을 확인해보세요. 이전에 했던 피드백들도 다시 한 번 떠올려보세요.

AppleCEO committed:
생성 - regexGrammar enum type 생성

정규표현식 문법을 저장할 파일 생성

AppleCEO committed:
수정 - 정규표현식 문자열로 하드코딩하지 않고 enum 타입에서 가져오도록 수정

AppleCEO committed:
수정 - 실수로 띄어놓은 줄들 제거

AppleCEO committed:
수정 - makeJson 메서드명을 jsonParse 로 변경

의미가 불명확하다는 피드백을 받아서 메서드명을 더 명확하게 변경

AppleCEO committed:
수정 - 오타 수정

AppleCEO committed:
수정 - 오타 수정

AppleCEO committed:
수정 - 괄호 시작과 끝에 rawValue 로 비교하던 것을 타입으로 비교하도록 수정

타입을 만든 것이니 직접 비교해보라는 피드백을 받고 수정

AppleCEO committed:
수정 - GrammarChecker 구조체 내 메서드명 개선

좀 더 명확한 의미를 가지는 메서드명으로 수정

AppleCEO committed:
수정 : inputToJson 메서드 이름을 stringToJson 으로 변경

AppleCEO committed:
생성 - JSONParserTest 테스트 생성

AppleCEO committed:
구현 - 테스트 일부 작성

AppleCEO committed:
수정 - 메서드명 수정

jsonParse 보다 parseJson 가 좀 더 맞는 것 같아서 메서드명 수정

AppleCEO committed:
구현 - 테스트 구현 완료

일부 버그 제거
테스트 정확하게 나오는 것 확인

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -43,43 +36,48 @@ struct GrammarChecker {
+        let arrayRegexGrammar = regexGrammar.array
+        if input.matches(arrayRegexGrammar.rawValue) {
             return true
         }
```
이런 비교문은 생략하고 그냥 return 에 결과를 바로 써도 되지 않을까요?

AppleCEO replied:
아.. 놓쳤습니다. 감사합니다!

godrm commented on JSONParser/JSONParserTest/JSONParserTest.swift:
```diff
@@ -0,0 +1,29 @@
+        XCTAssertNoThrow(try checkInputError("[ { \"name\" : \"dominic\" } ]"))
+        XCTAssertThrowsError(try checkInputError("[ \"name\" : \"KIM JUNG\" }"))
+        XCTAssertThrowsError(try checkInputError("[ { \"name\" : \"KIM JUNG\", \"alias\" : \"JK\", \"level\" : 5 }, { \"name\" : \"YAGOM\" } "))
+    }
```
이전에도 피드백을 했었지만 단위 테스트 하나의 경우에 대해 테스트하도록 구분해야 합니다. 그렇지 않으면 이중에서 어디서 실패했는지 알수가 없습니다.

AppleCEO replied:
하나의 메서드(GrammarChecker.checkJsonGrammar)에 대한 테스트로 수정하였습니다.

godrm replied:
하나의 메소드에 대한 단위 테스트라도 단위 테스트 함수 하나가 하나의 경우에 대해서만 테스트하는 게 좋습니다.
그래서 보통 단위 테스트 함수 내부에는 Assert 가 하나만 나오는 경우가 많습니다.

godrm commented on JSONParser/JSONParser/regexGrammar.swift:
```diff
@@ -0,0 +1,14 @@
+
+enum regexGrammar: String {
+    case object = "\\{[A-z\"0-9:., ]+\\}"
+    case array = "\\[[A-z\"0-9., ]+\\]"
```
여기 정규표현식에는 string, number, bool 에 대한 표현이 없는데도 예외처리도 되고 안정적으로 처리가 되나요?

AppleCEO replied:
예외처리가 안되고 있었습니다. 표현정규식을 공부해서 stiring, number, bool 에 대해서 안정적으로 처리되도록 수정하였습니다.

AppleCEO committed:
수정 - 의미없는 if 문에 따른 return 삭제

return 에 바로 if 문을 넣으면 되는 상황이라는 피드백을 받고 깨우침

AppleCEO committed:
수정 - object 정규표현식을 정확하게 수정

AppleCEO committed:
수정 - 표현정규식 개선

AppleCEO committed:
제거 - 표현정규식으로 한번에 객체와 배열을 검증할 수 있게 되면서 나눠서  각각 검증하던 메서드 제거

AppleCEO committed:
수정 - Conver.stringToJson 에 try 를 제거

표현정규식으로 공백이 없는지 까지 검증할 수 있게 되면서 기존에 문자열을 자르면서도 오류 처리하던 부분을 제거
테스트 소스에서도 stringToJson 을 제거

AppleCEO committed:
수정 - StringToJson 은 에러를 뱉지않도록 수정하였기 때문에 while 문 밖으로 위치 변경

AppleCEO committed:
수정 - 정규표현식을 통한 검증으로 수정

만든 정규표현식을 활용하지 않고 있어서 기존 검증하던 소스를 제거하고 정규표현식을 통해 검증하도록 구현

AppleCEO committed:
수정 - 정규표현식 버그 수정

\가 실수로 하나 더 들어갔는지 테스트가 제대로 되지않아서 다시 작성

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -21,71 +21,28 @@ extension String {
 
 struct GrammarChecker {
     static func checkJsonGrammar(_ input: String) throws {
```
저는 checkJsonGrammar() 함수가 return 값이 있고 throws 없는게 좋을 것 같습니다. 이 부분도 한 번 고민해보세요.
